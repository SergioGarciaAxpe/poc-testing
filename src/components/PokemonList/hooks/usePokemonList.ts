import { useMemo, useState } from 'react'
import { getPokemonList } from '../helpers/getPokemonList';
import { PokemonData } from '../types/PokemonList.types';

export const usePokemonList = () => {
    const [pokemonList, setPokemonList] = useState<PokemonData[]>([]);
    const [offset, setOffset] = useState(0);
  
    useMemo(async () => {
      const pokemonList = await getPokemonList(offset);
      setPokemonList(pokemonList);
    }, [offset]);
  
    const onPrevPage = () => setOffset(offset - 20 < 0 ? 0 : offset - 20);
  
    const onNextPage = () => setOffset(offset + 20 > 1281 ? 1281 : offset + 20);

    return {
        offset,
        pokemonList,
        onNextPage,
        onPrevPage
    }
}
