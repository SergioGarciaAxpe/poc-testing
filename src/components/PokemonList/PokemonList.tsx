import { usePokemonList } from "./hooks/usePokemonList";

export const PokemonList = () => {
  const { pokemonList, onNextPage, onPrevPage, offset } = usePokemonList();

  return (
    <div>
      {pokemonList.map((pokemonItem) => {
        return <div key={pokemonItem.name}>{JSON.stringify(pokemonItem)}</div>;
      })}

      <div style={{ margin: 20 }}>
        Showing from {offset + 1} to {offset + 20}
      </div>

      <div className="button-row">
        <button onClick={onPrevPage}>Prev</button>
        <button onClick={onNextPage}>Next</button>
      </div>
    </div>
  );
};
