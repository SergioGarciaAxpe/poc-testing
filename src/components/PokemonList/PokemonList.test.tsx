import { render, screen, fireEvent, renderHook } from "@testing-library/react";
import "@testing-library/jest-dom";
import { PokemonList } from "./PokemonList";
import { usePokemonList } from "./hooks/usePokemonList";

describe("PokemonList", () => {
  const mockPokemonList = [
    { name: "bulbasaur", url: "https://pokeapi.co/api/v2/pokemon/1/" },
    { name: "charmander", url: "https://pokeapi.co/api/v2/pokemon/4/" },
    { name: "squirtle", url: "https://pokeapi.co/api/v2/pokemon/7/" },
  ];

  test("renders the list of Pokemon", () => {
    const mockUsePokemonList = () => ({
      pokemonList: mockPokemonList,
      onNextPage: jest.fn(),
      onPrevPage: jest.fn(),
      offset: 0,
    });
    jest.mock("./hooks/usePokemonList", () => ({
      usePokemonList: mockUsePokemonList,
    }));

    const { asFragment } = render(<PokemonList />);

    expect(asFragment()).toMatchSnapshot(`
      <div>{"name":"bulbasaur","url":"https://pokeapi.co/api/v2/pokemon/1/"}</div>
      <div>{"name":"charmander","url":"https://pokeapi.co/api/v2/pokemon/4/"}</div>
      <div>{"name":"squirtle","url":"https://pokeapi.co/api/v2/pokemon/7/"}</div>
    `);
  });

  test("displays the correct range of Pokemon", () => {
    const { result } = renderHook(() => usePokemonList());
    const { offset, pokemonList } = result.current;

    expect(offset).toBe(0)
    expect(pokemonList.length).toBe(0)
    render(<PokemonList />);

    const rangeText = screen.getByText('Showing from 1 to 20');
    expect(rangeText).toBeInTheDocument();
    
    fireEvent.click(screen.getByText('Next'))

    const newRangeText = screen.getByText('Showing from 21 to 40');
    expect(newRangeText).toBeInTheDocument();
  });
});
