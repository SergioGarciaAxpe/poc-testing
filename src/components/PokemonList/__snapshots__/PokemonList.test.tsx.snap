// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonList renders the list of Pokemon: 
      <div>{"name":"bulbasaur","url":"https://pokeapi.co/api/v2/pokemon/1/"}</div>
      <div>{"name":"charmander","url":"https://pokeapi.co/api/v2/pokemon/4/"}</div>
      <div>{"name":"squirtle","url":"https://pokeapi.co/api/v2/pokemon/7/"}</div>
     1`] = `
<DocumentFragment>
  <div>
    <div
      style="margin: 20px;"
    >
      Showing from 1 to 20
    </div>
    <div
      class="button-row"
    >
      <button>
        Prev
      </button>
      <button>
        Next
      </button>
    </div>
  </div>
</DocumentFragment>
`;
